Userland/ntpquery.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2020, Nico Weber <thakis@chromium.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE
+#include <LibCore/ArgsParser.h>
+#include <arpa/inet.h>
+#include <endian.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+
+struct [[gnu::packed]] NtpPacket {
+    uint8_t li_vn_mode;
+    uint8_t stratum;
+    int8_t poll;
+    int8_t precision;
+
+    uint32_t root_delay;
+    uint32_t root_dispersion;
+    uint32_t reference_id;
+
+    uint64_t reference_timestamp;
+    uint64_t origin_timestamp;
+    uint64_t receive_timestamp;
+    uint64_t transmit_timestamp;
+};
+
+static_assert(sizeof(NtpPacket) == 48);
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio inet dns", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    // FIXME: Request serenityos.pool.ntp.org here https://manage.ntppool.org/manage/vendor
+    // and then use that as the default value.
+    // Until then, explicitly pass this as `ntpquery pool.ntp.org`.
+    const char* host = nullptr;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(host, "NTP server", "host", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    if (!host)
+        return 1;
```

Comment:
Not cool, make it required then

Reply:
See the FIXME 10 lines further up.

Reply:
Yes, but while we have no default value, and have to explicitly pass a host, it's not OK to just silently `exit(1)` without outputting an error message to the user (and not even a single line to the debug log!).

Just mark it as `Core::ArgsParser::Required::Yes`, and `ArgsParser` will do everything (ensure it's not-null) for you. When `serenityos.pool.ntp.org` becomes available, we put it in as a default value and switch the argument back to being not required.

Reply:
@nico Nor here

Reply:
Again: Done, but spending time on getting temporary code to be Just So doesn't seem useful (nor fun) to me.